### PR TITLE
Added "propr" snippet for required properties

### LINKED ISSUE #75128: Create Required Property in C# with propr
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Create-Required-Property-in-C-with-prop/10722495)._

---
In C#, I create property using the shortcut 'prop'. I want to be able to create required property using 'propr'

---
#### Feedback Bot on 8/14/2024, 07:52 PM: 

(private comment, text removed)

### PR DESCRIPTION
**Pull request by victor-pogor**
closes #75128

I have checked what modifiers are allowed and limited the `propr` snippet to the valid cases:

```c#
class C
{
    private required int Prop1 { get; set; } // Not valid
    protected required int Prop2 { get; set; } // Not valid
    internal required int Prop3 { get; set; } // OK
    private protected required int Prop4 { get; set; } // Not valid
    protected internal required int Prop5 { get; set; } // OK
    public required int Prop6 { get; set; } // OK
}

interface I
{
    required int Prop { get; set; } // Not valid
}
```

https://github.com/user-attachments/assets/2caeff90-9277-48db-90e5-fbd662a43813

### CODE REVIEW

File: src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf [hunk 1]
```diff
@@ -4200,6 +4200,11 @@ Chcete pokračovat?</target>
         <target state="translated">přístupový objekt události</target>
         <note />
       </trans-unit>
+      <trans-unit id="required_property">
+        <source>required property</source>
+        <target state="new">požadovaná vlastnost</target>
```
Comment (victor-pogor) [inline src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf:4205]: I've used the machine translation and can't be sure about the accuracy.

File: src/Features/Core/Portable/xlf/FeaturesResources.ru.xlf [hunk 2]
```diff
@@ -4200,6 +4200,11 @@ Do you want to continue?</source>
         <target state="translated">метод доступа к событию</target>
         <note />
       </trans-unit>
+      <trans-unit id="required_property">
+        <source>required property</source>
+        <target state="new">Обязательное свойство</target>
```
Comment (victor-pogor) [inline src/Features/Core/Portable/xlf/FeaturesResources.ru.xlf:4205]: should be the lowercase naming? I was referring to `property_` localization key which starts with the capital letter

File: src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs [hunk 3]
```diff
@@ -143,25 +143,13 @@ public Program()
             """);
     }
 
-    [Theory]
-    [MemberData(nameof(CommonSnippetTestData.AllAccessibilityModifiers), MemberType = typeof(CommonSnippetTestData))]
-    public async Task InsertSnippetAfterAccessibilityModifierTest(string modifier)
```
Comment (victor-pogor) [inline src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs:148]: I have moved out this test in the derived classes, not all modifiers are valid for required properties:

image

File: src/Features/Core/Portable/Snippets/CommonSnippetIdentifiers.cs [hunk 4]
```diff
@@ -13,5 +13,6 @@ internal static class CommonSnippetIdentifiers
     public const string ConsoleWriteLine = "cw";
     public const string Constructor = "ctor";
     public const string Property = "prop";
+    public const string RequiredProperty = "propr";
```
Comment (DoctorKrolic) [inline src/Features/Core/Portable/Snippets/CommonSnippetIdentifiers.cs:16]: "Required proeprty" is a C#-specific snippet (same as `propi` for instance), so please move this to `CSharpSnippetIdentifiers`

File: src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs (hunk 3 above)
Comment (DoctorKrolic) [reply src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs:148]: I would split this test into 2 parts with different modifiers, make the second part `virtual` and override it in test class for `propr` snippet. Your solution requires duplicating code and it is generally not the best approach

File: src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            readonly partial struct MyStruct
+            {
+            }
+            """, "public required {|0:int|} {|1:MyProperty|} { get; }");
+    }
+
+    [WorkItem("https://github.com/dotnet/roslyn/issues/75128")]
+    public override async Task InsertSnippetInInterfaceTest()
+    {
+        await VerifyDefaultPropertyAsync("""
```
Comment (DoctorKrolic) [inline src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs:68]: [Required properties are not valid in interfaces](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfWBDLCAY1QAIBJAWACgBveiligAQGYL1UBHAVyg8AJhUIUACugD2ABwoMKAc1QIA3BQDOqjQF96uoA=)

File: src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf (hunk 1 above)
Comment (DoctorKrolic) [reply src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf:4205]: You don't need to translate anything. The localization team will take care of that. You just add the item in English and that's it

File: src/Features/Core/Portable/xlf/FeaturesResources.ru.xlf (hunk 2 above)
Comment (DoctorKrolic) [reply src/Features/Core/Portable/xlf/FeaturesResources.ru.xlf:4205]: In general translation should keep the style of the original message. Since original message starts with lowercase, so should the translation. But again, this is a task for a localization team

File: src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf (hunk 1 above)
Comment (victor-pogor) [reply src/Features/Core/Portable/xlf/FeaturesResources.cs.xlf]: reverted the localizations to English

File: src/Features/Core/Portable/xlf/FeaturesResources.ru.xlf (hunk 2 above)
Comment (victor-pogor) [reply src/Features/Core/Portable/xlf/FeaturesResources.ru.xlf]: reverted the localizations to English

File: src/Features/Core/Portable/Snippets/CommonSnippetIdentifiers.cs (hunk 4 above)
Comment (victor-pogor) [reply src/Features/Core/Portable/Snippets/CommonSnippetIdentifiers.cs]: Good catch, moved it out

File: src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs (hunk 5 above)
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs]: Updated the test

File: src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs (hunk 3 above)
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs:148]: I did some changes, please check

File: src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                {{modifier}} $$
+            }
+            """, $$"""required {|0:int|} {|1:MyProperty|} {{DefaultPropertyBlockText}}""");
+    }
+
+    [Theory, WorkItem("https://github.com/dotnet/roslyn/issues/75128")]
+    [InlineData("private")]
+    [InlineData("protected")]
+    [InlineData("private protected")]
+    public async Task DoNotInsertSnippetAfterAccessibilityModifierTest(string modifier)
```
[the remark this answers is not in the document]
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs]: good point, renamed the tests

File: src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            readonly partial struct MyStruct
+            {
+            }
+            """, "public required {|0:int|} {|1:MyProperty|} { get; }");
+    }
+
+    [WorkItem("https://github.com/dotnet/roslyn/issues/75128")]
+    public override async Task InsertSnippetInInterfaceTest()
```
[the remark this answers is not in the document]
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs]: renamed

File: src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs [hunk 8]
```diff
@@ -67,6 +67,17 @@ record MyRecord
             """);
     }
 
+    [Fact]
+    public async Task InsertSnippetInRecordStructTest()
```
[the remark this answers is not in the document]
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs]: Merged the test cases into one test

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpAutoPropertySnippetProvider.cs [hunk 9]
```diff
@@ -87,4 +89,9 @@ protected override ImmutableArray<SnippetPlaceholder> GetPlaceHolderLocationsLis
         var node = root.FindNode(TextSpan.FromBounds(position, position));
         return node.GetAncestorOrThis<PropertyDeclarationSyntax>();
     }
+
+    protected virtual SyntaxToken[] GetAdditionalPropertyModifiers(CSharpSyntaxContext? syntaxContext)
+    {
+        return [];
```
[the remark this answers is not in the document]
Comment (victor-pogor) [reply src/Features/CSharp/Portable/Snippets/AbstractCSharpAutoPropertySnippetProvider.cs]: Agree, fixed

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpAutoPropertySnippetProvider.cs [hunk 10]
```diff
@@ -87,4 +89,9 @@ protected override ImmutableArray<SnippetPlaceholder> GetPlaceHolderLocationsLis
         var node = root.FindNode(TextSpan.FromBounds(position, position));
         return node.GetAncestorOrThis<PropertyDeclarationSyntax>();
     }
+
+    protected virtual SyntaxToken[] GetAdditionalPropertyModifiers(CSharpSyntaxContext? syntaxContext)
```
[the remark this answers is not in the document]
Comment (victor-pogor) [reply src/Features/CSharp/Portable/Snippets/AbstractCSharpAutoPropertySnippetProvider.cs]: moved

File: src/Features/CSharp/Portable/Snippets/CSharpProprSnippetProvider.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+        // The required modifier can't be applied to members of an interface
+        if (syntaxContext.ContainingTypeDeclaration is InterfaceDeclarationSyntax)
+            return false;
+
+        if (syntaxContext.PrecedingModifiers.IsEmpty())
+            return true;
+
+        // "private" and "private protected" modifiers are NOT valid for required property
+        if (precedingModifiers.Any(syntaxKind => syntaxKind == SyntaxKind.PrivateKeyword))
```
[the remark this answers is not in the document]
Comment (victor-pogor) [reply src/Features/CSharp/Portable/Snippets/CSharpProprSnippetProvider.cs]: Good one, this helped to reduce one `if` condition 😉

File: src/Features/CSharp/Portable/Snippets/CSharpProprSnippetProvider.cs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+        var syntaxContext = (CSharpSyntaxContext)context.SyntaxContext;
+        var precedingModifiers = syntaxContext.PrecedingModifiers;
+
+        // The required modifier can't be applied to members of an interface
+        if (syntaxContext.ContainingTypeDeclaration is InterfaceDeclarationSyntax)
+            return false;
+
+        if (syntaxContext.PrecedingModifiers.IsEmpty())
+            return true;
```
Comment (DoctorKrolic) [inline src/Features/CSharp/Portable/Snippets/CSharpProprSnippetProvider.cs:44]: If I see it correctly this `if` is also not needed. All subsequent checks won't trigger, so we will get to the end and return true there

File: src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs [hunk 13]
```diff
@@ -56,11 +56,14 @@ class MyClass
             """);
     }
 
-    [Fact]
-    public async Task InsertSnippetInRecordTest()
+    [Theory]
+    [InlineData("record")]
+    [InlineData("record struct")]
+    [InlineData("record class")]
+    public async Task InsertSnippetInRecordTest(string text)
```
Comment (DoctorKrolic) [inline src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs:63]: Please some more descriptive name for the parameter. I suggest `recordType`, but if you come up with something better, use that instead

File: src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Snippets;
+
+public sealed class CSharpProprSnippetProviderTests : AbstractCSharpAutoPropertySnippetProviderTests
+{
+    protected override string SnippetIdentifier => "propr";
+
+    protected override string DefaultPropertyBlockText => "{ get; set; }";
+
+    [WorkItem("https://github.com/dotnet/roslyn/issues/75128")]
```
Comment (DoctorKrolic) [inline src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs:18]: I don't see much value in adding `WorkItem` to every test when you implement a feature from scratch. They are useful when fixing bugs or extending the feature to see how it evolved over time and what edge cases were discovered, but having them like that... not so much. Like what you expect to be done when new tests are added? They will technically be related to the feature request, so is the attribute needed to be duped on every test? I don't like that idea. Consider removing them. Other property snippet tests don't have `WorkItem`s all over the place either

File: src/Features/CSharp/Portable/Snippets/CSharpProprSnippetProvider.cs (hunk 12 above)
Comment (victor-pogor) [reply src/Features/CSharp/Portable/Snippets/CSharpProprSnippetProvider.cs]: Yeah, you're right. Removed.

File: src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs (hunk 13 above)
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/AbstractCSharpAutoPropertySnippetProviderTests.cs]: Renamed the param name

File: src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs (hunk 14 above)
Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/CSharpProprSnippetProviderTests.cs]: Not sure what are the practices in Roslyn for new features, happy to remove the `WorkItem` attributes though

File: src/Features/CSharpTest/Snippets/CSharpPropSnippetProviderTests.cs [hunk 15]
```diff
@@ -65,4 +66,8 @@ interface MyInterface
             }
             """);
     }
+
+    [Theory]
+    [MemberData(nameof(CommonSnippetTestData.AllAccessibilityModifiers), MemberType = typeof(CommonSnippetTestData))]
+    public override Task InsertSnippetAfterAllowedAccessibilityModifierTest(string modifier) => base.InsertSnippetAfterAllowedAccessibilityModifierTest(modifier);
```
Comment (DoctorKrolic) [inline src/Features/CSharpTest/Snippets/CSharpPropSnippetProviderTests.cs:72]: Please add a line break like that for all these tests. This is _too_ long line. It was just not that obvious from side-by-side diff
```suggestion
    public override Task InsertSnippetAfterAllowedAccessibilityModifierTest(string modifier)
        => base.InsertSnippetAfterAllowedAccessibilityModifierTest(modifier);
```

Comment (victor-pogor) [reply src/Features/CSharpTest/Snippets/CSharpPropSnippetProviderTests.cs]: fixed

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 16]
```diff
@@ -3171,4 +3171,7 @@ Zero-width positive lookbehind assertions are typically used at the beginning of
   <data name="Symbol_search" xml:space="preserve">
     <value>Symbol search</value>
   </data>
+  <data name="required_property" xml:space="preserve">
+    <value>required property</value>
```
Comment (DoctorKrolic) [inline src/Features/Core/Portable/FeaturesResources.resx:3175]: Since "required property" is a C#-only concept, consider moving this to `CSharpFeaturesResources` as well.

I apologize for the fact, that I'm producing only a few comments at a time instead of leaving all of them at once. I am not an experienced reviewer, so my eye catches things gradually over time

Comment (victor-pogor) [reply src/Features/Core/Portable/FeaturesResources.resx]: Good point, moved to `CSharpFeaturesResources`

**Review by DoctorKrolic — APPROVED** [review]
LGTM. My opinion doesn't count really, but I expect a lot less feedback from the team member now. @victor-pogor You may resolve all leftover feedback items from me. Thanks for great work!

**Comment by DoctorKrolic** [comment]
@akhera99 PTAL

**Review by akhera99 — APPROVED** [review]
Looks good, thank you @victor-pogor for the contribution and @DoctorKrolic for reviewing!
